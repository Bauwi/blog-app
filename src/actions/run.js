import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
