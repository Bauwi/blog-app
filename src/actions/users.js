import { db } from '../firebase/firebase';

export const updateUser = (id, updates) => ({
  type: 'UPDATE_USER',
  id,
  updates
});

export const startUpdateUser = (id, updates) => (dispatch, getState) => {
  const uid = id || getState().auth.uid;
  return db
    .collection('users')
    .doc(uid)
    .get()
    .then((doc) => {
      if (doc.exists) {
        return db
          .collection('users')
          .doc(uid)
          .update(updates)
          .then(() => {
            dispatch(updateUser(uid, updates));
          });
      }
      return db
        .collection('users')
        .doc(uid)
        .set({
          avatar: 'http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png',
          email: '',
          username: '',
          stars: 0,
          topCategories: ['life', 'programming', 'Music']
        })
        .then(startUpdateUser(uid, updates));
    });
};

export const addUserStar = nextStars => ({
  type: 'ADD_USER_STAR',
  nextStars
});

export const startAddUserStar = (id, prevStars) => dispatch =>
  db
    .collection('users')
    .doc(id)
    .update({ stars: prevStars + 1 })
    .then(() => dispatch(addUserStar(prevStars + 1)));

export const setAuthor = author => ({
  type: 'SET_AUTHOR',
  author
});

export const startSetAuthor = id => dispatch =>
  db
    .collection('users')
    .doc(id)
    .get()
    .then((doc) => {
      dispatch(setAuthor(doc.data()));
    });

export const setUserPreferences = preferences => ({
  type: 'SET_USER_PREFERENCES',
  preferences
});

export const startSetUserPreferences = preferences => (dispatch, getState) =>
  db
    .collection('users')
    .doc(getState().auth.uid)
    .get()
    .then((doc) => {
      dispatch(setUserPreferences(doc.data()));
    });
