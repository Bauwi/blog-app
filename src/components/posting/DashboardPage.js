import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import PostsList from './PostsList';
import PostsSummary from './PostsSummary';
import UserCard from '../UserCard';

const DashboardPage = ({ preferences, userId }) => (
  <div>
    <PostsSummary />

    <div className="content-container">
      <UserCard author={preferences} userId={userId} />
      <PostsList />
    </div>
  </div>
);

DashboardPage.propTypes = {
  preferences: PropTypes.shape({
    avatar: PropTypes.string,
    description: PropTypes.string,
    email: PropTypes.string,
    stars: PropTypes.number.isRequired,
    topCategories: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string
  }).isRequired,
  userId: PropTypes.string.isRequired
};

const mapStateToProps = state => ({
  preferences: state.users.preferences,
  userId: state.auth.uid
});

export { DashboardPage };
export default connect(mapStateToProps)(DashboardPage);
