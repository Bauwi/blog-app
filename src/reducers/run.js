const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
