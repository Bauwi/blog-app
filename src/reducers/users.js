const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_USER':
      return {
        ...state,
        preferences: action.newUser
      };
    case 'UPDATE_USER_PREFERENCES':
      return {
        ...state,
        preferences: { ...state.preferences, ...action.updates }
      };
    case 'SET_AUTHOR':
      return {
        ...state,
        author: action.author
      };
    case 'ADD_USER_STAR':
      return {
        ...state,
        author: {
          ...state.author,
          stars: action.nextStars
        }
      };
    case 'SET_USER_PREFERENCES':
      return {
        ...state,
        preferences: action.preferences
      };
    case 'USERS_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'USERS_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    default:
      return state;
  }
};

export default usersReducer;
