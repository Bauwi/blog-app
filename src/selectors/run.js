const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
