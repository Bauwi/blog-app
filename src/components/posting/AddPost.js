import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PostForm from './PostForm';
import { startAddPost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.startUpdateUser(undefined, { username: post.author });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <header className="content-container">
            <Link to="/dashboard">Back to dashboard</Link>
          </header>
        </div>
        <PostForm onSubmit={this.onSubmit} context="add" author={this.props.user} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

const mapStateToProps = state => ({
  user: state.users.preferences
});
export default connect(mapStateToProps, mapDispatchToProps)(AddPost);
