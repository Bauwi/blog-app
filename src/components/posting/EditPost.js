import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startEditPost, startRemovePost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

import PostForm from './PostForm';

export class EditPost extends Component {
  onSubmit = post => {
    this.props.startEditPost(this.props.post.id, post);
    this.props.history.push('/dashboard');
  };
  onRemovePost = () => {
    this.props.startRemovePost(this.props.post.id);
    this.props.startUpdateUser(undefined, {
      numberOfPosts: this.props.preferences.numberOfPosts - 1
    });
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <PostForm
          onSubmit={this.onSubmit}
          post={this.props.post}
          author={this.props.preferences}
          context="edit"
        />
        <div className="content-container">
          <button className="button" onClick={this.onRemovePost}>
            Remove Post
          </button>
        </div>
      </div>
    );
  }
}

EditPost.propTypes = {
  post: PropTypes.shape({
    id: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    authorId: PropTypes.string.isRequired,
    body: PropTypes.arrayOf(PropTypes.object).isRequired,
    cover: PropTypes.string.isRequired,
    createdAt: PropTypes.number.isRequired,
    keywords: PropTypes.string.isRequired,
    miniCover: PropTypes.string.isRequired,
    readingTime: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired
  }).isRequired,
  preferences: PropTypes.shape({
    avatar: PropTypes.string,
    description: PropTypes.string,
    email: PropTypes.string,
    stars: PropTypes.number.isRequired,
    topCategories: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string
  }).isRequired,
  startEditPost: PropTypes.func.isRequired,
  startRemovePost: PropTypes.func.isRequired
};

const mapStateToProps = (state, props) => {
  return {
    post: state.posts.posts.find(post => {
      return post.id === props.match.params.id;
    }),
    preferences: state.users.preferences
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: id => dispatch(startRemovePost(id)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);
