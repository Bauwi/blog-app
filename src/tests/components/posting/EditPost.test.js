import React from 'react';
import { shallow } from 'enzyme';
import { EditPost } from '../../../components/posting/EditPost';
import run from '../../fixtures/run';
import { user1 } from '../../fixtures/users';

let startEditPost;
let startUpdateUser;
let startRemovePost;
let history;
let wrapper;

beforeEach(() => {
  startEditPost = jest.fn();
  startUpdateUser = jest.fn();
  startRemovePost = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditPost
    post={run.posts[2].content}
    startEditPost={startEditPost}
    startUpdateUser={startUpdateUser}
    startRemovePost={startRemovePost}
    history={history}
    preferences={user1}
  />);
});

test('should render EditPost component properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onRemove', () => {
  wrapper.find('button').simulate('click');
  expect(startRemovePost).toHaveBeenCalled();
  expect(history.push).toHaveBeenLastCalledWith('/dashboard');
});

test('should handle onSubmit edit post', () => {
  wrapper.find('PostForm').prop('onSubmit')(run.posts[2]);
  expect(startEditPost).toHaveBeenLastCalledWith(run.posts[2].content.id, run.posts[2]);
  expect(history.push).toHaveBeenLastCalledWith('/dashboard');
});
