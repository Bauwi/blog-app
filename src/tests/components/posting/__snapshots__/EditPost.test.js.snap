// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPost component properly 1`] = `
<div>
  <PostForm
    author={
      Object {
        "avatar": "http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png",
        "description": "I had three chairs in my house; one for solitude, two for friendship, three for society.",
        "email": "",
        "numberOfPosts": 3,
        "stars": 12,
        "topCategories": Array [
          "music",
          "movie",
          "litterature",
        ],
        "username": "user1",
      }
    }
    context="edit"
    onSubmit={[Function]}
    post={
      Object {
        "author": "author3",
        "authorId": "author3Id",
        "body": Array [
          Object {
            "attributes": "h3",
            "text": "blabla",
          },
        ],
        "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
        "createdAt": 1000000,
        "id": "3",
        "keywords": "car, pool, cake",
        "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
        "readingTime": 2,
        "stars": 2,
        "title": "Post 3 in Run",
      }
    }
  />
  <div
    className="content-container"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Remove Post
    </button>
  </div>
</div>
`;
